stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/colwise_reduction.hpp>
+#include <stan/math/opencl/kernel_generator/rowwise_reduction.hpp>
+#include <stan/math/opencl/kernel_generator/calc_if.hpp>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace stan {
+namespace math {
+/** \addtogroup opencl_kernel_generator
+ *  @{
+ */
+
+namespace internal {
+class reduction_2d_base {};
+}  // namespace internal
+
+/**
+ * Represents a two dimensional reduction in kernel generator expressions. So as
+ * to be efficient two dimensional reductions are only done partially. That
+ * means instead of 1 element kernel output can have a few rows and a few
+ * columns that need to be reduced to obtain final result. This can be done in a
+ * separate kernel or after copying to CPU. Also two dimensional reductions can
+ * not be used as arguments to other operations - they can only be evaluated.
+ * @tparam Derived derived type
+ * @tparam T type of first argument
+ * @tparam Operation type with member function generate that accepts two
+ * variable names and returns OpenCL source code for reduction operation_cl
+ */
+template <typename Derived, typename T, typename Operation>
+class reduction_2d
+    : public internal::reduction_2d_base,
+      public operation_cl<Derived, typename std::remove_reference_t<T>::Scalar,
+                          T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl<Derived, Scalar, T>;
+  using base::var_name_;
+  static const bool require_specific_local_size = true;
+
+ protected:
+  std::string init_;
+  using base::derived;
+
+ public:
+  using base::cols;
+  /**
+   * Constructor
+   * @param a the expression to reduce
+   * @param init OpenCL source code of initialization value for reduction
+   */
+  explicit reduction_2d(T&& a, const std::string& init)
+      : base(std::forward<T>(a)), init_(init) {}
+
+  /**
+   * Generates kernel code for assigning this expression into result expression.
+   * @param[in,out] generated map from (pointer to) already generated local
+   * operations to variable names
+   * @param[in,out] generated_all map from (pointer to) already generated all
+   * operations to variable names
+   * @param ng name generator for this kernel
+   * @param row_index_name row index variable name
+   * @param col_index_name column index variable name
+   * @param result expression into which result is to be assigned
+   * @return part of kernel with code for this and nested expressions
+   */
+  template <typename T_result>
+  kernel_parts get_whole_kernel_parts(
+      std::map<const void*, const char*>& generated,
+      std::map<const void*, const char*>& generated_all, name_generator& ng,
```

Comment:
We are putting in here pointers to both operations and `matrix_cl`s (see load operation), so it can not be `operation_cl*`. These objects can be const so the pointer needs to be const (alternatively we could const cast it, but that is ugly). Also I don't understand how exactly you think we could use string for this, but I am sure it would be inefficient.

Reply:
> We are putting in here pointers to both operations and matrix_cls (see load operation), so it can not be operation_cl*. These objects can be const so the pointer needs to be const (alternatively we could const cast it, but that is ugly)

I'm fine with the const. If we need both what to do these `void*` get cast to eventually or how is that decided? For me here it's mostly an issue of readability of passing in a void* pointer. Like could we just have an empty base class both operations and matric_cl inherit from? Or a 


> Also I don't understand how exactly you think we could use string for this, but I am sure it would be inefficient.

nvm about this it was a different issue

https://github.com/stan-dev/math/issues/1297

Reply:
> get cast to eventually or how is that decided?

No. We never need to cast these pointers to anything else. I don't understand the second part, How is what decided?

> Like could we just have an empty base class both operations and matric_cl inherit from?

Adding a base class just for this seems quite a bit forced. I had an idea of merging both of these classes, but I think each of them is large enough as it is.

Reply:
If @rok-cesnovar is fine with the void then I'll be okay approving this

Reply:
Yeah, I am good with that.